diff --git a/fla/ops/gla/chunk.py b/fla/ops/gla/chunk.py
index 0ea9f330..0663381e 100644
--- a/fla/ops/gla/chunk.py
+++ b/fla/ops/gla/chunk.py
@@ -321,6 +321,7 @@ def chunk_gla_fwd_kernel_o(
     chunk_indices,
     scale,
     T,
+    scalar,
     H: tl.constexpr,
     HV: tl.constexpr,
     K: tl.constexpr,
@@ -376,9 +377,9 @@ def chunk_gla_fwd_kernel_o(
         b_h = tl.load(p_h, boundary_check=(0, 1))
         if i_k >= 0:
             if TRANSPOSE_STATE:
-                b_o += tl.dot(b_qg, tl.trans(b_h).to(b_qg.dtype))
+                b_o += tl.dot(b_qg, tl.trans(b_h).to(b_qg.dtype)) * scalar
             else:
-                b_o += tl.dot(b_qg, b_h.to(b_qg.dtype))
+                b_o += tl.dot(b_qg, b_h.to(b_qg.dtype)) * scalar
     b_o *= scale
     p_v = tl.make_block_ptr(v, (T, V), (HV*V, 1), (i_t * BT, i_v * BV), (BT, BV), (1, 0))
     p_o = tl.make_block_ptr(o, (T, V), (HV*V, 1), (i_t * BT, i_v * BV), (BT, BV), (1, 0))
@@ -388,7 +389,7 @@ def chunk_gla_fwd_kernel_o(
     # [BT, BT]
     b_A = tl.load(p_A, boundary_check=(0, 1))
     b_A = tl.where(m_s, b_A, 0.).to(b_v.dtype)
-    b_o += tl.dot(b_A, b_v)
+    b_o += tl.dot(b_A, b_v) * scalar
     tl.store(p_o, b_o.to(p_o.dtype.element_ty), boundary_check=(0, 1))
 
 
@@ -896,6 +897,7 @@ def chunk_gla_fwd_o_gk(
 
     # Please ensure zeros, since vllm will use padding v
     o = torch.zeros_like(v)
+    scalar = 1.0
     def grid(meta): return (triton.cdiv(V, meta['BV']), NT, B * HV)
     chunk_gla_fwd_kernel_o[grid](
         q=q,
@@ -913,6 +915,7 @@ def chunk_gla_fwd_o_gk(
         K=K,
         V=V,
         BT=BT,
+        scalar=scalar,
         USE_EXP2=use_exp2,
         TRANSPOSE_STATE=transpose_state_layout,
     )
diff --git a/fla/ops/kda/chunk_bwd.py b/fla/ops/kda/chunk_bwd.py
index 68bfa0c1..f80e013b 100644
--- a/fla/ops/kda/chunk_bwd.py
+++ b/fla/ops/kda/chunk_bwd.py
@@ -21,9 +21,9 @@ from fla.ops.utils.constant import RCP_LN2
 from fla.ops.utils.op import exp2
 from fla.utils import IS_NVIDIA_HOPPER, autotune_cache_kwargs, check_shared_mem
 
-BK_LIST = [32, 64] if check_shared_mem() else [16, 32]
-BV_LIST = [64, 128] if check_shared_mem('ampere') else [16, 32]
-NUM_WARPS = [2, 4] if IS_NVIDIA_HOPPER else [2, 4, 8]
+BK_LIST = [64] if check_shared_mem() else [32]
+BV_LIST = [64] if check_shared_mem('ampere') else [32]
+NUM_WARPS = [2] if IS_NVIDIA_HOPPER else [2]
 
 
 @triton.heuristics({
@@ -144,6 +144,7 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
     chunk_indices,
     scale,
     T,
+    scalar,
     H: tl.constexpr,
     HV: tl.constexpr,
     K: tl.constexpr,
@@ -193,9 +194,11 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
 
     p_beta = tl.make_block_ptr(beta, (T,), (HV,), (i_t * BT,), (BT,), (0,))
     b_beta = tl.load(p_beta, boundary_check=(0,))
+    tl.extra.cann.extension.compile_hint(b_beta, "mayDiscretememaccess")
 
     p_A = tl.make_block_ptr(A, (BT, T), (1, HV * BT), (0, i_t * BT), (BT, BT), (0, 1))
     b_A = tl.load(p_A, boundary_check=(0, 1))
+    tl.extra.cann.extension.compile_hint(b_A, "mayDiscretememaccess")
 
     b_dA = tl.zeros([BT, BT], dtype=tl.float32)
     b_db = tl.zeros([BT], dtype=tl.float32)
@@ -208,9 +211,12 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
         p_g = tl.make_block_ptr(g, (T, K), (HV*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
         b_k = tl.load(p_k, boundary_check=(0, 1))
         b_g = tl.load(p_g, boundary_check=(0, 1)).to(tl.float32)
+        tl.extra.cann.extension.compile_hint(b_k, "mayDiscretememaccess")
+        tl.extra.cann.extension.compile_hint(b_g, "mayDiscretememaccess")
 
         p_gn = g + (min(T, i_t * BT + BT) - 1).to(tl.int64) * HV*K + o_k
         b_gn = tl.load(p_gn, mask=m_k, other=0).to(tl.float32)
+        tl.extra.cann.extension.compile_hint(b_gn, "mayDiscretememaccess")
 
         b_dq = tl.zeros([BT, BK], dtype=tl.float32)
         b_dk = tl.zeros([BT, BK], dtype=tl.float32)
@@ -230,30 +236,39 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
             # [BT, BV]
             b_v_new = tl.load(p_v_new, boundary_check=(0, 1))
             b_do = tl.load(p_do, boundary_check=(0, 1))
+            tl.extra.cann.extension.compile_hint(b_v_new, "mayDiscretememaccess")
+            tl.extra.cann.extension.compile_hint(b_do, "mayDiscretememaccess")
             # [BV, BK]
             b_h = tl.load(p_h, boundary_check=(0, 1))
             b_dh = tl.load(p_dh, boundary_check=(0, 1))
+            tl.extra.cann.extension.compile_hint(b_h, "mayDiscretememaccess")
+            tl.extra.cann.extension.compile_hint(b_dh, "mayDiscretememaccess")
             # [BT, BV]
             b_dv = tl.load(p_dv, boundary_check=(0, 1))
+            tl.extra.cann.extension.compile_hint(b_dv, "mayDiscretememaccess")
 
             b_dgk += tl.sum(b_h * b_dh, axis=0)
-            b_dq += tl.dot(b_do, b_h.to(b_do.dtype))
-            b_dk += tl.dot(b_v_new, b_dh.to(b_v_new.dtype))
-            b_dw += tl.dot(b_dv.to(b_v_new.dtype), b_h.to(b_v_new.dtype))
+            b_dq += tl.dot(b_do, b_h.to(b_do.dtype))  * scalar
+            b_dk += tl.dot(b_v_new, b_dh.to(b_v_new.dtype))  * scalar
+            b_dw += tl.dot(b_dv.to(b_v_new.dtype), b_h.to(b_v_new.dtype))  * scalar
             tl.debug_barrier()  # DO NOT REMOVE THIS LINE!
             if i_k == 0:
                 p_v = tl.make_block_ptr(v, (T, V), (HV*V, 1), (i_t * BT, i_v * BV), (BT, BV), (1, 0))
                 p_dv2 = tl.make_block_ptr(dv2, (T, V), (HV*V, 1), (i_t * BT, i_v * BV), (BT, BV), (1, 0))
 
                 b_v = tl.load(p_v, boundary_check=(0, 1))
+                tl.extra.cann.extension.compile_hint(b_v, "mayDiscretememaccess")
 
-                b_dA += tl.dot(b_dv, tl.trans(b_v))
+                b_dA += tl.dot(b_dv, tl.trans(b_v)) * scalar
 
                 b_dvb = tl.dot(b_A, b_dv)
                 b_dv2 = b_dvb * b_beta[:, None]
                 b_db += tl.sum(b_dvb * b_v, 1)
 
-                tl.store(p_dv2, b_dv2.to(p_dv2.dtype.element_ty), boundary_check=(0, 1))
+                casted_b_dv2 = b_dv2.to(p_dv2.dtype.element_ty)
+                tl.extra.cann.extension.compile_hint(casted_b_dv2, "mayDiscretememaccess")
+                tl.store(p_dv2, casted_b_dv2, boundary_check=(0, 1))
+ 
 
         b_gk_exp = exp2(b_g)
         b_gb = b_gk_exp * b_beta[:, None]
@@ -264,13 +279,14 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
         b_kg = b_k * b_gk_exp
 
         b_dw = -b_dw.to(b_A.dtype)
-        b_dA += tl.dot(b_dw, tl.trans(b_kg.to(b_A.dtype)))
+        b_dA += tl.dot(b_dw, tl.trans(b_kg.to(b_A.dtype))) * scalar
 
         b_dkgb = tl.dot(b_A, b_dw)
         b_db += tl.sum(b_dkgb * b_kg, 1)
 
         p_q = tl.make_block_ptr(q, (T, K), (H*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
         b_q = tl.load(p_q, boundary_check=(0, 1))
+        tl.extra.cann.extension.compile_hint(b_q, "mayDiscretememaccess")
         b_kdk = b_k * b_dk
         b_dgk += tl.sum(b_kdk, axis=0)
         b_dg = b_q * b_dq - b_kdk + m_last[:, None] * b_dgk + b_kg * b_dkgb * b_beta[:, None]
@@ -279,9 +295,15 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
         p_dq = tl.make_block_ptr(dq, (T, K), (HV*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
         p_dk = tl.make_block_ptr(dk, (T, K), (HV*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
         p_dg = tl.make_block_ptr(dg, (T, K), (HV*K, 1), (i_t * BT, i_k * BK), (BT, BK), (1, 0))
-        tl.store(p_dq, b_dq.to(p_dq.dtype.element_ty), boundary_check=(0, 1))
-        tl.store(p_dk, b_dk.to(p_dk.dtype.element_ty), boundary_check=(0, 1))
-        tl.store(p_dg, b_dg.to(p_dg.dtype.element_ty), boundary_check=(0, 1))
+        casted_b_dq = b_dq.to(p_dq.dtype.element_ty)
+        tl.extra.cann.extension.compile_hint(casted_b_dq, "mayDiscretememaccess")
+        tl.store(p_dq, casted_b_dq, boundary_check=(0, 1))
+        casted_b_dk = b_dk.to(p_dk.dtype.element_ty)
+        tl.extra.cann.extension.compile_hint(casted_b_dk, "mayDiscretememaccess")
+        tl.store(p_dk, casted_b_dk, boundary_check=(0, 1))
+        casted_b_dg = b_dg.to(p_dg.dtype.element_ty)
+        tl.extra.cann.extension.compile_hint(casted_b_dg, "mayDiscretememaccess")
+        tl.store(p_dg, casted_b_dg, boundary_check=(0, 1))
 
     m_A = (o_t[:, None] > o_t[None, :]) & (m_t[:, None] & m_t)
     b_dA = tl.where(m_A, b_dA * b_beta[None, :], 0)
@@ -291,8 +313,12 @@ def chunk_kda_bwd_kernel_wy_dqkg_fused(
 
     p_dA = tl.make_block_ptr(dA, (T, BT), (HV * BT, 1), (i_t * BT, 0), (BT, BT), (1, 0))
     p_db = tl.make_block_ptr(db, (T,), (HV,), (i_t * BT,), (BT,), (0,))
-    tl.store(p_dA, b_dA.to(p_dA.dtype.element_ty), boundary_check=(0, 1))
-    tl.store(p_db, b_db.to(p_db.dtype.element_ty), boundary_check=(0,))
+    casted_b_dA = b_dA.to(p_dA.dtype.element_ty)
+    tl.extra.cann.extension.compile_hint(casted_b_dA, "mayDiscretememaccess")
+    tl.store(p_dA, casted_b_dA, boundary_check=(0, 1))
+    casted_b_db = b_db.to(p_db.dtype.element_ty)
+    tl.extra.cann.extension.compile_hint(casted_b_db, "mayDiscretememaccess")
+    tl.store(p_db, casted_b_db, boundary_check=(0,))
 
 
 def chunk_kda_bwd_dAv(
@@ -380,6 +406,7 @@ def chunk_kda_bwd_wy_dqkg_fused(
     dg = torch.empty_like(g, dtype=torch.float)
     db = torch.empty_like(beta, dtype=torch.float)
     dA = torch.empty_like(A, dtype=torch.float)
+    scalar = 1.0
 
     grid = (NT, B * HV)
     chunk_kda_bwd_kernel_wy_dqkg_fused[grid](
@@ -404,6 +431,7 @@ def chunk_kda_bwd_wy_dqkg_fused(
         chunk_indices=chunk_indices,
         scale=scale,
         T=T,
+        scalar=scalar,
         H=H,
         HV=HV,
         K=K,
diff --git a/fla/ops/kda/chunk_intra.py b/fla/ops/kda/chunk_intra.py
index df0ba02a..9a7b5851 100644
--- a/fla/ops/kda/chunk_intra.py
+++ b/fla/ops/kda/chunk_intra.py
@@ -31,8 +31,8 @@ else:
 @triton.autotune(
     configs=[
         triton.Config({'BK': BK}, num_warps=num_warps)
-        for BK in [32, 64]
-        for num_warps in [1, 2, 4]
+        for BK in [64] # 32, 64
+        for num_warps in [1] # 1, 2, 4
     ],
     key=["H", "HV", "K", "BC"],
     **autotune_cache_kwargs,
@@ -50,6 +50,7 @@ def chunk_kda_fwd_kernel_inter_solve_fused(
     cu_seqlens,
     chunk_indices,
     T,
+    scalar,
     H: tl.constexpr,
     HV: tl.constexpr,
     K: tl.constexpr,
@@ -144,8 +145,8 @@ def chunk_kda_fwd_kernel_inter_solve_fused(
             # [BK, BC]
             b_kgt = tl.trans(b_k0 * exp2(b_gn1[None, :] - b_g0))
             # [BC, BC]
-            b_Aqk10 += tl.dot(b_q1 * b_gqn, b_kgt)
-            b_Akk10 += tl.dot(b_k1 * b_gqn, b_kgt)
+            b_Aqk10 += tl.dot(b_q1 * b_gqn, b_kgt) * scalar
+            b_Akk10 += tl.dot(b_k1 * b_gqn, b_kgt) * scalar
 
             if i_tc2 < T:
                 p_q2 = tl.make_block_ptr(q, (T, K), (H*K, 1), (i_tc2, i_k * BK), (BC, BK), (1, 0))
@@ -163,13 +164,13 @@ def chunk_kda_fwd_kernel_inter_solve_fused(
                 b_kg2 = b_k2 * b_gqn2
                 # [BK, BC]
                 b_kgt = tl.trans(b_k0 * exp2(b_gn2[None, :] - b_g0))
-                b_Aqk20 += tl.dot(b_qg2, b_kgt)
-                b_Akk20 += tl.dot(b_kg2, b_kgt)
+                b_Aqk20 += tl.dot(b_qg2, b_kgt) * scalar
+                b_Akk20 += tl.dot(b_kg2, b_kgt) * scalar
                 # [BC, BC]
                 b_kgt = tl.trans(b_k1 * exp2(b_gn2[None, :] - b_g1))
                 # [BC, BC]
-                b_Aqk21 += tl.dot(b_qg2, b_kgt)
-                b_Akk21 += tl.dot(b_kg2, b_kgt)
+                b_Aqk21 += tl.dot(b_qg2, b_kgt) * scalar
+                b_Akk21 += tl.dot(b_kg2, b_kgt) * scalar
 
                 if i_tc3 < T:
                     p_q3 = tl.make_block_ptr(q, (T, K), (H*K, 1), (i_tc3, i_k * BK), (BC, BK), (1, 0))
@@ -188,18 +189,18 @@ def chunk_kda_fwd_kernel_inter_solve_fused(
                     # [BK, BC]
                     b_kgt = tl.trans(b_k0 * exp2(b_gn3[None, :] - b_g0))
                     # [BC, BC]
-                    b_Aqk30 += tl.dot(b_qg3, b_kgt)
-                    b_Akk30 += tl.dot(b_kg3, b_kgt)
+                    b_Aqk30 += tl.dot(b_qg3, b_kgt) * scalar
+                    b_Akk30 += tl.dot(b_kg3, b_kgt) * scalar
                     # [BK, BC]
                     b_kgt = tl.trans(b_k1 * exp2(b_gn3[None, :] - b_g1))
                     # [BC, BC]
-                    b_Aqk31 += tl.dot(b_qg3, b_kgt)
-                    b_Akk31 += tl.dot(b_kg3, b_kgt)
+                    b_Aqk31 += tl.dot(b_qg3, b_kgt) * scalar
+                    b_Akk31 += tl.dot(b_kg3, b_kgt) * scalar
                     # [BK, BC]
                     b_kgt = tl.trans(b_k2 * exp2(b_gn3[None, :] - b_g2))
                     # [BC, BC]
-                    b_Aqk32 += tl.dot(b_qg3, b_kgt)
-                    b_Akk32 += tl.dot(b_kg3, b_kgt)
+                    b_Aqk32 += tl.dot(b_qg3, b_kgt) * scalar
+                    b_Akk32 += tl.dot(b_kg3, b_kgt) * scalar
 
     ################################################################################
     # save off-diagonal Aqk blocks and prepare Akk
@@ -814,6 +815,7 @@ def chunk_kda_fwd_intra(
 
     # Step 2: Fused inter + solve_tril (works for both fixed-len and varlen)
     grid = (NT, B * HV)
+    scalar = 1.0
     chunk_kda_fwd_kernel_inter_solve_fused[grid](
         q=q,
         k=k,
@@ -826,12 +828,14 @@ def chunk_kda_fwd_intra(
         cu_seqlens=cu_seqlens,
         chunk_indices=chunk_indices,
         T=T,
+        scalar=scalar,
         H=H,
         HV=HV,
         K=K,
         BT=BT,
         BC=BC,
         USE_SAFE_GATE=safe_gate,
+        inject_barrier_all=True,
     )
     w, u, qg, kg = recompute_w_u_fwd(
         k=k,
